rama-http/src/matcher/subdomain_trie.rs

```diff
@@ -0,0 +1,173 @@
+            }
+        }
+        
+        false
+    }
+
+}
+
+fn reverse_domain(domain: &str) -> String {
+    let from = domain.strip_prefix('.').unwrap_or(domain);
+    let mut domain = from.split('.').rev().collect::<Vec<&str>>().join(".");
+    domain.push('.');
+    domain
+}
+
+impl<State, Body> Matcher<State, Request<Body>> for SubdomainTrieMatcher {
+    fn matches(
+        &self,
+        ext: Option<&mut Extensions>,
+        ctx: &Context<State>,
+        req: &Request<Body>,
+    ) -> bool {
+        let host = match ctx.get::<RequestContext>() {
+            Some(req_ctx) => req_ctx.authority.host().clone(),
+            None => {
+                let req_ctx: RequestContext = match (ctx, req).try_into() {
+                    Ok(rc) => rc,
+                    Err(err) => {
+                        tracing::error!(error = %err, "SubdomainTrieMatcher: failed to extract request context");
+                        return false;
+                    }
+                };
+                let host = req_ctx.authority.host().clone();
+                if let Some(ext) = ext {
+                    ext.insert(req_ctx);
+                }
+                host
+            }
+        };
+
+        match host {
+            Host::Name(domain) => {
+                let is_match = self.is_match(domain.as_str());
+                tracing::trace!("SubdomainTrieMatcher: matching '{}' => {}", domain, is_match);
+                is_match
+            }
+            Host::Address(_) => {
+                tracing::trace!("SubdomainTrieMatcher: ignoring numeric address");
+                false
+            }
+        }
+    }
+}
+
+impl FromIterator<String> for SubdomainTrieMatcher {
+    fn from_iter<I: IntoIterator<Item = String>>(iter: I) -> Self {
+        let domains: Vec<String> = iter.into_iter().collect();
+
+        SubdomainTrieMatcher::new(domains)
+    }
+}
+
+impl<'a> FromIterator<&'a str> for SubdomainTrieMatcher {
+    fn from_iter<I: IntoIterator<Item = &'a str>>(iter: I) -> Self {
+        let domains: Vec<&str> = iter.into_iter().collect();
+
+        SubdomainTrieMatcher::new(domains)
+    }
+}
+
+#[cfg(test)]
+mod subdomain_trie_tests {
+    use super::*;
+
+    #[test]
+    fn test_reverse_domain() {
+        assert_eq!(reverse_domain("example.com"), "com.example.");
+        assert_eq!(reverse_domain(".example.com"), "com.example.");
+        assert_eq!(reverse_domain("sub.example.com"), "com.example.sub.");
+    }
+
+    #[test]
+    fn test_trie_matching() {
+        let matcher = SubdomainTrieMatcher::new(vec!["example.com", "sub.domain.org"]);
+        assert!(matcher.is_match("example.com"));
+        assert!(matcher.is_match(".example.com"));
+        assert!(matcher.is_match("sub.domain.org"));
+        assert!(matcher.is_match("sub.example.com"));
+        assert!(!matcher.is_match("domain.org"));
+        assert!(!matcher.is_match("other.com"));
+    }
+
+    #[test]
+    fn test_from_iterator_string() {
```

Comment:
just need one of these as they are all the same, and you can just do it with a `["example.com", "sub.domain.org"]` kind of test